rfcs/react-components/convergence/motion-definition-n-apis.md

```diff
@@ -0,0 +1,999 @@
+        <MyButton />
+      </Fade>
+    </>
+  );
+}
+```
+
+To make it work in this scenario, `MyButton` should handle the `className` prop:
+
+```tsx
+function MyButton() {
+  return <Button className={props.className}>Hello world!</Button>;
+}
+```
+
+The issue is that there is no way to enforce it in the API contract.
+
+</details>
+
+> Note: both CSS and Web Animations API options require to pass refs to elements, but that's easier to enforce as we check that a child uses `React.forwardRef()`.
+
+## Detailed design
+
+The proposed solution relies on the [Web Animations API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Animations_API) and proposes to use the factories pattern to create React components for animating elements.
+
+<details>
+<summary>Why factories over hooks?</summary>
+
+To execute animations, we need a motion definition to create an animation. With hooks, this definition can be defined also inside a component, as illustrated below:
+
+```tsx
+// ⚠️ This is not proposed API, it's just an example
+
+function MyComponent() {
+  const motionRef = useAtomMotion({
+    keyframes: {
+      from: { opacity: 0 },
+      to: { opacity: 1 },
+    },
+    duration: 2000,
+  });
+
+  return <div ref={motionRef}>Hello world!</div>;
+}
+```
+
+This approach presents a design issue: on every render, a new definition object is created and passed to the hook. At the same time, the motion must be a dependency to start the animation. Consequently, a new animation is created and started on every render, as depicted in the example below:
+
+```tsx
+function useAtomMotion(motion) {
+  const elementRef = React.useRef();
+
+  React.useEffect(() => {
+    const animation = elementRef.current.animate(motion.keyframes, {
+      duration: motion.duration,
+    });
+
+    return () => {
+      animation.cancel();
+    };
+  }, [motion]);
+}
+```
+
+With the factories pattern, the same problem is not possible, as a motion is defined outside a component, as shown below:
+
+```tsx
+function createAtomMotion(motion) {
+  return React.forwardRef((props, ref) => {
+    const elementRef = React.useRef();
+
+    React.useEffect(() => {
+      elementRef.current = elementRef.current.animate(motion.keyframes, {
+        duration: motion.duration,
+      });
+
+      return () => {
+        elementRef.current.cancel();
+      };
+      // `motion` does not need to be a dependency as it's defined outside the component
+    }, []);
+
+    return React.cloneElement(props.children, { ref: elementRef });
+  });
+}
+```
+
+</details>
+
+### Motion definition
+
+As the Web Animations API is framework-agnostic, we don't need to use Griffel to generate CSS keyframes. Instead, we can define them in JS:
+
+```tsx
+const fadeEnterSlow: MotionAtom = {
+  keyframes: {
+    from: { opacity: 0 },
+    to: { opacity: 1 },
+  },
+  fill: 'forwards',
+
+  // Heads up! `duration` is in milliseconds and `easing` is a string
+  // ⚠️ We can't use CSS variables there
```

Comment:
**Disclaimer**: I am not leaving this comment as a recommendation for what is best for MSFT, but talking more in general terms 😄 

With moving to CSS variables + static styles extraction, I expect more and more theme definitions to depend on CSS variables - including here the tokens related to animations. Also, with the rose of RSC I would suppose that you would want to limit the requirement for using context (unless you are opting for the server context) to a minimum (in case you want these motion tokens to be defined via some kind of context). This is my only worry with this kind of API.

Tagging @brijeshb42 and @siriwatknp in case they want to give their thoughts on this RFC.

Reply:
@mnajdova that's a good callout. The original explanation was not complete and confusing 😕 I added a new section that explains CSS vars usage in 6dd795c 👍 

TL;DR CSS variables could be used in [`keyframes`](https://react.fluentui.dev/?path=/docs/utilities-web-motions-preview--motion-default#tokens-usage) as is. The limitation appears only for [`options`](https://developer.mozilla.org/en-US/docs/Web/API/KeyframeEffect/KeyframeEffect#options), but as motions could be defined as factories - there is a workaround for it.

Reply:
> The limitation appears only for options, but as motions could be defined as factories - there is a workaround for it.

Thanks for the explanation, it's good to know that it can be solved in a way. However, I suppose this means that the component would need to re-render for the change to be applied no? You would have to force re-render if a CSS var changes, while with CSS animations that won't be the case. Anyway, this feature may not be in the requirements you have, so maybe it doesn't matter that much.

Reply:
> However, I suppose this means that the component would need to re-render for the change to be applied no? You would have to force re-render if a CSS var changes, while with CSS animations that won't be the case.

@mnajdova yeah, you're correct about this. However, even when a value changes for a variable it will not trigger a new animation 🤔